### PR TITLE
Log provider: fix DB issues and enable to update filter

### PR DESCRIPTION
**Pull request by amirylm**

### CODE REVIEW

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go [hunk 1]
```diff
@@ -30,6 +30,9 @@ var (
 
 	// AllowedLogsPerUpkeep is the maximum number of logs allowed per upkeep every single call.
 	AllowedLogsPerUpkeep = 5
+
+	readJobQueueSize = 64
+	readLogsTimeout  = 5 * time.Second
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go:35]: would 5 seconds be enough for upkeeps with a lot of logs?

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go [hunk 2]
```diff
@@ -188,7 +191,10 @@ func (p *logEventProvider) GetLatestPayloads(context.Context) ([]ocr2keepers.Upk
 }
 
 // ReadLogs fetches the logs for the given upkeeps.
-func (p *logEventProvider) ReadLogs(ctx context.Context, force bool, ids ...*big.Int) error {
+func (p *logEventProvider) ReadLogs(pctx context.Context, force bool, ids ...*big.Int) error {
+	ctx, cancel := context.WithTimeout(pctx, readLogsTimeout)
+	defer cancel()
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go:196]: not able to comment on the exact line, if this context cancels then we would get an error in line 204.
I see we then update filters in line 205, should the error handling (line 207) be done before that?

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go (hunk 1 above)
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go:35]: Increasing to 10sec

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go (hunk 2 above)
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go:205]: There might be an error for some of the filters, therefore we always try to update last poll block of those that are have managed to read successfully

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider_life_cycle.go [hunk 3]
```diff
@@ -17,30 +17,62 @@ var (
...
 	}
-	filter := p.newLogFilter(upkeepID, cfg)
+	lpFilter := p.newLogFilter(upkeepID, cfg)
 
-	if p.filterStore.Has(upkeepID) {
-		// TODO: check if filter was updated, if so unregister previous filter and register new one
-		return errors.Errorf("filter for upkeep with id %s already registered", upkeepID.String())
+	// using lock to facilitate multiple events causing filter registration
+	// at the same time.
+	// TODO: consider using a q to handle registration requests
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider_life_cycle.go:29]: nit. not sure if you are planning to do this before launch, if not can you update to exploratory

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go [hunk 4]
```diff
@@ -13,6 +13,7 @@ import (
 	"github.com/ethereum/go-ethereum/common"
 	coreTypes "github.com/ethereum/go-ethereum/core/types"
 	"github.com/pkg/errors"
+	"go.uber.org/multierr"
```
Comment (FelixFan1992) [inline core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:16]: just nit: can we use multi error in go native library?

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/integration_test.go [hunk 5]
```diff
@@ -107,6 +107,18 @@ func TestIntegration_LogEventProvider(t *testing.T) {
 		}()
 		defer logProvider2.Close()
 
+		// re-register filters
+		for i, id := range ids {
+			err = logProvider2.RegisterFilter(logprovider.FilterOptions{
+				UpkeepID:      id,
+				TriggerConfig: newPlainLogTriggerConfig(addrs[i]),
```
Comment (ferglor) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/integration_test.go:114]: Will `addrs` always be the same size as `ids`?

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go [hunk 6]
```diff
@@ -341,10 +363,13 @@ func (p *logEventProvider) readLogs(ctx context.Context, latest int64, filters [
 		if len(filter.addr) == 0 {
 			continue
 		}
-		// start should either be the last block polled for the filter or the
-		// lookback range in the case this is the first time the filter is polled
+		// first time polling, use filter.configUpdateBlock as a starting point of the range
 		start := filter.lastPollBlock
-		if start == 0 || start < latest-lookbackBlocks {
+		if start == 0 {
+			start = int64(filter.configUpdateBlock)
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go:369]: We should have this condition after line 387 where we apply reorg buffer

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go [hunk 7]
```diff
@@ -83,13 +96,14 @@ func NewLogProvider(lggr logger.Logger, poller logpoller.LogPoller, packer LogDa
...
-		buffer:      newLogEventBuffer(lggr, int(opts.LookbackBlocks), BufferMaxBlockSize, AllowedLogsPerBlock),
-		poller:      poller,
-		lock:        sync.RWMutex{},
-		opts:        opts,
-		filterStore: filterStore,
+		packer:       packer,
+		lggr:         lggr.Named("KeepersRegistry.LogEventProvider"),
+		buffer:       newLogEventBuffer(lggr, int(opts.LookbackBlocks), BufferMaxBlockSize, AllowedLogsPerBlock),
+		poller:       poller,
+		lock:         sync.RWMutex{},
```
Comment (FelixFan1992) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go:103]: just nit: no need to init mutexes

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider_life_cycle.go [hunk 8]
```diff
@@ -17,42 +18,117 @@ var (
...
+			if err := p.UnregisterFilter(id); err != nil {
+				merr = errors.Join(merr, fmt.Errorf("failed to unregister filter: %s", id.String()))
+			}
+		}
+	}
+	var newIDs []*big.Int
+	for id, ok := range visited {
+		if !ok {
+			uid, ok := new(big.Int).SetString(id, 10)
+			if !ok {
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider_life_cycle.go:49]: nit. not sure if overriding `ok` here would have any repercussions

File: core/services/ocr2/plugins/ocr2keeper/integration_21_test.go [hunk 9]
```diff
@@ -226,18 +226,26 @@ func TestIntegration_KeeperPluginLogUpkeep(t *testing.T) {
 	g.Eventually(listener, testutils.WaitTimeout(t), cltest.DBPollingInterval).Should(gomega.BeTrue())
 	done()
 
-	runs := checkPipelineRuns(t, nodes, 1*len(nodes)) // TODO: TBD
+	runs := checkPipelineRuns(t, nodes, 1) // TODO: TBD
```
Comment (ferglor) [inline core/services/ocr2/plugins/ocr2keeper/integration_21_test.go:229]: Has this TODO been fixed or is it still TBD even after this change?

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go [hunk 10]
```diff
@@ -270,22 +271,39 @@ func (r *EvmRegistry) refreshActiveUpkeeps() error {
...
-	// for any upkeeps that might have been removed. We depend upon state events to remove the configs
+	return r.refreshLogTriggerUpkeeps(ids)
+}
+
+// refreshLogTriggerUpkeeps refreshes the active upkeep ids for log trigger upkeeps
+//
+// TODO: manage reorgs, one idea is to pull the last ConfigSet event for each upkeep,
+// and to call updateTriggerConfig with latest config and block number.
+// Calling the log provider with newer block will update the filter accordingly.
+func (r *EvmRegistry) refreshLogTriggerUpkeeps(ids []*big.Int) error {
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:282]: I am not fully following the behaviour here. Want to confirm

- It removes filters for upkeeps that do not exist anymore
- For existing upkeeps it ignores them?  (Not sure how we can manage reorgs then?)
- For new upkeeps that somehow didn't have a filter, adds them with creation block 0 (should we use current latest block here?)

Comment (infiloop2) [reply core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:282]: `one idea is to pull the last ConfigSet event for each upkeep`

I realised we would also need to look for unpaused event

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider_life_cycle.go [hunk 11]
```diff
@@ -17,42 +18,113 @@ var (
...
 
 func (p *logEventProvider) UnregisterFilter(upkeepID *big.Int) error {
 	err := p.poller.UnregisterFilter(p.filterName(upkeepID))
-	if err == nil {
-		p.filterStore.RemoveActiveUpkeeps(upkeepFilter{
-			upkeepID: upkeepID,
-		})
+	if err != nil {
+		// TODO: mark as removed in filter store, so we'll
+		// automatically retry on next refresh
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider_life_cycle.go:121]: this would automatically be handled now by our RefreshActiveUpkeeps logic upon next refresh?

File: core/services/ocr2/plugins/ocr2keeper/integration_21_test.go [hunk 12]
```diff
@@ -230,14 +230,22 @@ func TestIntegration_KeeperPluginLogUpkeep(t *testing.T) {
...
 		// blockBeforeEmits := backend.Blockchain().CurrentBlock().Number.Uint64()
 		// Emit 100 logs in a burst
 		emits := 100
-		emitEvents(testutils.Context(t), t, 100, []*log_upkeep_counter_wrapper.LogUpkeepCounter{contract}, carrol, func() {})
+		i := 0
+		emitEvents(testutils.Context(t), t, 100, []*log_upkeep_counter_wrapper.LogUpkeepCounter{contract}, carrol, func() {
+			i++
+			if i%(emits/4) == 0 {
+				backend.Commit()
+				time.Sleep(time.Millisecond * 250) // otherwise we get "invalid transaction nonce" errors
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/integration_21_test.go:246]: i think adding constant time values are going to add to flakiness, but can be resolved in another PR

**Review by infiloop2 — APPROVED** [review]
